libraries/AP_Scripting/drivers/BattMon_ANX.lua

```diff
@@ -0,0 +1,264 @@
+   0xFF9F, 0xEFBE, 0xDFDD, 0xCFFC, 0xBF1B, 0xAF3A, 0x9F59, 0x8F78,
+   0x9188, 0x81A9, 0xB1CA, 0xA1EB, 0xD10C, 0xC12D, 0xF14E, 0xE16F,
+   0x1080, 0x00A1, 0x30C2, 0x20E3, 0x5004, 0x4025, 0x7046, 0x6067,
+   0x83B9, 0x9398, 0xA3FB, 0xB3DA, 0xC33D, 0xD31C, 0xE37F, 0xF35E,
+   0x02B1, 0x1290, 0x22F3, 0x32D2, 0x4235, 0x5214, 0x6277, 0x7256,
+   0xB5EA, 0xA5CB, 0x95A8, 0x8589, 0xF56E, 0xE54F, 0xD52C, 0xC50D,
+   0x34E2, 0x24C3, 0x14A0, 0x0481, 0x7466, 0x6447, 0x5424, 0x4405,
+   0xA7DB, 0xB7FA, 0x8799, 0x97B8, 0xE75F, 0xF77E, 0xC71D, 0xD73C,
+   0x26D3, 0x36F2, 0x0691, 0x16B0, 0x6657, 0x7676, 0x4615, 0x5634,
+   0xD94C, 0xC96D, 0xF90E, 0xE92F, 0x99C8, 0x89E9, 0xB98A, 0xA9AB,
+   0x5844, 0x4865, 0x7806, 0x6827, 0x18C0, 0x08E1, 0x3882, 0x28A3,
+   0xCB7D, 0xDB5C, 0xEB3F, 0xFB1E, 0x8BF9, 0x9BD8, 0xABBB, 0xBB9A,
+   0x4A75, 0x5A54, 0x6A37, 0x7A16, 0x0AF1, 0x1AD0, 0x2AB3, 0x3A92,
+   0xFD2E, 0xED0F, 0xDD6C, 0xCD4D, 0xBDAA, 0xAD8B, 0x9DE8, 0x8DC9,
+   0x7C26, 0x6C07, 0x5C64, 0x4C45, 0x3CA2, 0x2C83, 0x1CE0, 0x0CC1,
+   0xEF1F, 0xFF3E, 0xCF5D, 0xDF7C, 0xAF9B, 0xBFBA, 0x8FD9, 0x9FF8,
+   0x6E17, 0x7E36, 0x4E55, 0x5E74, 0x2E93, 0x3EB2, 0x0ED1, 0x1EF0
+}
+
+local function crc_ANX(bytes)
+   -- ANX CRC uses xmodem with a seed of 0xa635
+   local crc = 0xa635
+	for i=1,#bytes do
+		local b = string.byte(bytes,i,i)
+		crc = ((crc<<8) & 0xffff) ~ XMODEMCRC16Lookup[(((crc>>8)~b) & 0xff) + 1]
+	end
+    return crc
+end
+
+local frame_count = 0
+
+local function log_can_frame(frame)
+   logger.write("CANF",'Id,DLC,FC,B0,B1,B2,B3,B4,B5,B6,B7','IBIBBBBBBBB',
+                frame:id(),
+                frame:dlc(),
+                frame_count,
+                frame:data(0), frame:data(1), frame:data(2), frame:data(3),
+                frame:data(4), frame:data(5), frame:data(6), frame:data(7))
+   frame_count = frame_count + 1
+end
+ 
+
+local function parse_volt_frame(payload)
+   if #payload < 12 then
+      -- invalid length
+      return
+   end
+   local total_volt, current, rem_cap, temperature, _ = string.unpack("<HhHhH", string.sub(payload, 1, 10))
+   total_volt = total_volt * 0.01
+   -- it is a discharge, so use minus current
+   current = -current * 0.01
+   rem_cap = rem_cap * 0.1
+   temperature = temperature * 0.1
+   local num_cells = (#payload - 10) / 2
+   local cells = {}
+   for i = 1, num_cells do
+      cells[i] = string.unpack("<H", string.sub(payload, 11+(i-1)*2, 11+i*2))
+      cells[i] = cells[i]
+   end
+   -- API allows for up to 32 cells, but battery backend is more limited
+   if num_cells > 32 then
+      num_cells = 32
+   end
+
+   local state = BattMonitorScript_State()
+   state:healthy(true)
+   state:voltage(total_volt)
+   state:cell_count(num_cells)
+   state:capacity_remaining_pct(math.floor(rem_cap))
+   for i = 1, num_cells do
+      state:cell_voltages(i-1, cells[i])
+   end
+   state:cycle_count(255)
+   state:current_amps(current)
+   state:consumed_mah(NaN)
+   state:consumed_wh(NaN)
```

Comment:
Let's add a comment explaining that these `NaN` indicate that we want ArduPilot to calculate consumption by numerical integration, and not that we want ArduPilot to report that the consumption is unknown.

Reply:
In fact, let's put the comments explaining all 4 variables that can be set `NaN`, since the script is responsible for setting them.

- Temperature, set to `NaN` if unknown
- Current, set to `NaN` if unknown
- Consumption, set these to `NaN` to have them automatically calculated by numerical integration. Only set these manually if you have a better way of measuring them.

Reply:
yep, good idea, done